### PR TITLE
569 create option users

### PR DESCRIPTION
**Pull request by PaulRC-ioet**
A feature that inserts a new option to the administrator role to see the users.

### CODE REVIEW

File: src/app/modules/users/pages/users.component.html [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<div class="container">
+<app-users-list></app-users-list>
+</div>
```
Comment (Angeluz-07) [inline src/app/modules/users/pages/users.component.html:3]: It is missing a line at the end.

File: src/app/modules/users/models/users.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export interface User {
+  name: string;
+  email: string;
+  role?: string;
+  id: string;
+  tenand_id?: string;
```
Comment (Angeluz-07) [inline src/app/modules/users/models/users.ts:6]: It think there is a typo here, it should be `tenant_id`

Comment (Angeluz-07) [reply src/app/modules/users/models/users.ts:6]: Maybe all references to this property should be updated accordingly.

File: src/app/modules/users/components/users-list/users-list.component.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      .pipe(filter((action: any) => action.type === UserActionTypes.LOAD_USERS_SUCCESS))
+      .subscribe((action) => {
+        this.users = action.payload;
+        this.rerenderDataTable();
+      });
+    this.store.dispatch(new LoadUsers());
+  }
+
+  ngAfterViewInit(): void {
+    this.rerenderDataTable();
```
Comment (Angeluz-07) [inline src/app/modules/users/components/users-list/users-list.component.ts:37]: Only a doubt. It is ok to call the method twice?. Is called when the component is initialized and after the view is initialized.

File: src/app/modules/users/store/user.reducer.spec.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  const initialState: UserState = { data: [], isLoading: false, message: '' };
+
+  it('on LoadUser, isLoading is true', () => {
+    const action = new actions.LoadUsers();
+    const state = userReducer(initialState, action);
+
+    expect(state.isLoading).toEqual(true);
+  });
+
+  it('on LoadUserSucess, isLoading is flase and state has data', () => {
```
Comment (Angeluz-07) [inline src/app/modules/users/store/user.reducer.spec.ts:15]: Only a typo, `LoadUserSuccess` `false`

File: src/app/modules/users/store/user.reducer.spec.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import { UserState, userReducer } from './user.reducers';
+import { User } from '../models/users';
+import * as actions from './user.actions';
+
+describe('userReducer', () => {
+  const initialState: UserState = { data: [], isLoading: false, message: '' };
```
Comment (Angeluz-07) [inline src/app/modules/users/store/user.reducer.spec.ts:6]: We are not using the message property here, maybe we can remove it?

File: src/app/modules/users/store/user.effects.spec.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    effects = TestBed.inject(UserEffects);
+    service = TestBed.inject(UsersService);
+    toastrService = TestBed.inject(ToastrService);
+  });
+
+  it('should be created', async () => {
+    expect(effects).toBeTruthy();
+  });
+
+  it('should return a list of users when the action LOAD_USERS is called', async () => {
```
Comment (Angeluz-07) [inline src/app/modules/users/store/user.effects.spec.ts:33]: Maybe should be : 
`'should return a list of users when the action LOAD_USERS  call succeeds`

**Review by Angeluz-07 — COMMENTED** [review]
Good job. Only some comments.

File: src/app/modules/users/pages/users.component.html (hunk 1 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/pages/users.component.html:3]: Thanks, it's Done.

File: src/app/modules/users/models/users.ts (hunk 2 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/models/users.ts:6]: Sorry is a typo error, I have checked with the back end and I corrected already

File: src/app/modules/users/components/users-list/users-list.component.ts (hunk 3 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/components/users-list/users-list.component.ts:37]: It is not necessary for now, but it will be when the new functionalities are added.

File: src/app/modules/users/store/user.reducer.spec.ts (hunk 4 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/store/user.reducer.spec.ts:15]: Typo error, thanks!.

File: src/app/modules/users/store/user.effects.spec.ts (hunk 6 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/store/user.effects.spec.ts:33]: sounds better,  changed added thanks.

**Review by magallegos1996 — APPROVED** [review]

**Review by Angeluz-07 — APPROVED** [review]

File: src/app/modules/users/components/users-list/users-list.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+import { MockStore, provideMockStore } from '@ngrx/store/testing';
+
+import { NgxPaginationModule } from 'ngx-pagination';
+import { UsersListComponent } from './users-list.component';
+import { UserActionTypes, UserState, LoadUsers } from '../../store';
+import { ActionsSubject } from '@ngrx/store';
+import { DataTablesModule } from 'angular-datatables';
+import { act } from '@ngrx/effects';
```
Comment (enriquezrene) [inline src/app/modules/users/components/users-list/users-list.component.spec.ts:9]: unused import

File: src/app/modules/users/components/users-list/users-list.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+describe('UsersListComponent', () => {
+  let component: UsersListComponent;
+  let fixture: ComponentFixture<UsersListComponent>;
+  let store: MockStore<UserState>;
+  const actionSub: ActionsSubject = new ActionsSubject();
+
+  const state = {
```
Comment (enriquezrene) [inline src/app/modules/users/components/users-list/users-list.component.spec.ts:17]: please specify the type for this state

File: src/app/modules/users/components/users-list/users-list.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    data: [{ name: 'name', email: 'email', role: 'role', id: 'id', tenand_id: 'tenand id', deleted: 'delete' }],
+    isLoading: false,
+    message: '',
+  };
+
+  beforeEach(async(() => {
```
Comment (enriquezrene) [inline src/app/modules/users/components/users-list/users-list.component.spec.ts:23]: it is discouraged to keep using `async`, please use waitForAsync instead

File: src/app/modules/users/components/users-list/users-list.component.spec.ts (hunk 8 above)
Comment (enriquezrene) [reply src/app/modules/users/components/users-list/users-list.component.spec.ts:17]: this is UserState I guess so 
```
const state: UserState ....
```

File: src/app/modules/users/components/users-list/users-list.component.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
+import { AfterViewInit, Component, EventEmitter, Input, OnDestroy, OnInit, Output, ViewChild } from '@angular/core';
```
Comment (enriquezrene) [inline src/app/modules/users/components/users-list/users-list.component.ts:1]: EventEmitter, Input, and Output are not used imports

File: src/app/modules/users/pages/users.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    TestBed.configureTestingModule({
+      declarations: [ UsersComponent ]
+    })
+    .compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(UsersComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
```
Comment (enriquezrene) [inline src/app/modules/users/pages/users.component.spec.ts:19]: why is this needed?
```    
fixture.detectChanges();
```

File: src/app/modules/users/pages/users.component.ts [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+@Component({
+  selector: 'app-users',
+  templateUrl: './users.component.html',
+  styleUrls: ['./users.component.css']
+})
+export class UsersComponent implements OnInit {
+
+  constructor() { }
+
+  ngOnInit(): void {
```
Comment (enriquezrene) [inline src/app/modules/users/pages/users.component.ts:12]: what for? this component should not implement anything

File: src/app/modules/users/store/user.effects.spec.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+import { UserActionTypes } from './user.actions';
+import { UserEffects } from './user.effects';
+import { HttpClientTestingModule } from '@angular/common/http/testing';
+import { ToastrModule, ToastrService } from 'ngx-toastr';
+
+describe('UserEffects', () => {
+  let actions$: Observable<Action>;
+  let effects: UserEffects;
+  let service: UsersService;
+  let toastrService;
```
Comment (enriquezrene) [inline src/app/modules/users/store/user.effects.spec.ts:15]: why not defining the type here?
```
let toastrService: ToastrService
```

File: src/app/modules/users/store/user.effects.spec.ts [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    effects = TestBed.inject(UserEffects);
+    service = TestBed.inject(UsersService);
+    toastrService = TestBed.inject(ToastrService);
+  });
+
+  it('should be created', async () => {
+    expect(effects).toBeTruthy();
+  });
+
+  it('should return a list of users when the action LOAD_USERS call succeeds', async () => {
```
Comment (enriquezrene) [inline src/app/modules/users/store/user.effects.spec.ts:33]: The test name does not reflect what you're testing here. It should be:

```
 action type is LOAD_USER_SUCCESS when service is executed sucessfully 
```

File: src/app/modules/users/store/user.effects.spec.ts [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    actions$ = of({ type: UserActionTypes.LOAD_USERS });
+    const serviceSpy = spyOn(service, 'loadUsers');
+    serviceSpy.and.returnValue(of(user));
+
+    effects.loadUsers$.subscribe((action) => {
+      expect(action.type).toEqual(UserActionTypes.LOAD_USERS_SUCCESS);
+    });
+  });
+
+  it('should return an error when the action LOAD_USERS call fails', async () => {
```
Comment (enriquezrene) [inline src/app/modules/users/store/user.effects.spec.ts:43]: Wrong test description, you are not returning any error here. Instead, you're calling the error function from toastrService and you are expective a `LOAD_USER_FAIL` as action type

File: src/app/modules/users/store/user.reducer.spec.ts [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+  it('on LoadUser, isLoading is true', () => {
+    const action = new actions.LoadUsers();
+    const state = userReducer(initialState, action);
+
+    expect(state.isLoading).toEqual(true);
+  });
+
+  it('on LoadUserSucess, isLoading is false and state has data', () => {
```
Comment (enriquezrene) [inline src/app/modules/users/store/user.reducer.spec.ts:14]: typo Success not Sucess

**Review by enriquezrene — CHANGES_REQUESTED** [review]

**Comment by enriquezrene** [comment]
Also the table is hard to understand, 

what the last column means?
is there any chance to make the table wider?
Headers should be:
- Email
- Name
- Roles ---> I think a user might have more than one, right?
- idk, what it does?

image

**Comment by enriquezrene** [comment]
This also looks kinda ugly

image

File: src/app/modules/users/components/users-list/users-list.component.spec.ts (hunk 7 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/components/users-list/users-list.component.spec.ts:9]: I have already removed them, thanks

File: src/app/modules/users/components/users-list/users-list.component.spec.ts (hunk 8 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/components/users-list/users-list.component.spec.ts:17]: yes, I already added the type

File: src/app/modules/users/components/users-list/users-list.component.spec.ts (hunk 9 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/components/users-list/users-list.component.spec.ts:23]: in this version it does not let me use waitForAsync, I am waiting for Sandro's changes to add it and not have conflicts in .json file

File: src/app/modules/users/components/users-list/users-list.component.ts (hunk 10 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/components/users-list/users-list.component.ts:1]: I have already removed them, thanks

File: src/app/modules/users/pages/users.component.spec.ts (hunk 11 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/pages/users.component.spec.ts:19]: These are functions that come by default when generating a component in angular, all the files that are generated have that. should I remove them?

File: src/app/modules/users/pages/users.component.ts (hunk 12 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/pages/users.component.ts:12]: These are functions that come by default when generating a component in angular, but now I remove them

File: src/app/modules/users/store/user.effects.spec.ts (hunk 13 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/store/user.effects.spec.ts:15]: Done.

File: src/app/modules/users/store/user.effects.spec.ts (hunk 14 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/store/user.effects.spec.ts:33]: the description has been changed thanks

File: src/app/modules/users/store/user.effects.spec.ts (hunk 15 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/store/user.effects.spec.ts:43]: I have changed the description as in the previous comment to have a standard for the test effects.

File: src/app/modules/users/store/user.reducer.spec.ts (hunk 16 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/store/user.reducer.spec.ts:14]: Thanks, it's Done

**Comment by PaulRC-ioet** [comment]
Captura de pantalla de 2020-11-20 16-04-11

This is the new model of the main view of Users

File: src/app/modules/users/pages/users.component.spec.ts (hunk 11 above)
Comment (enriquezrene) [reply src/app/modules/users/pages/users.component.spec.ts:19]: If so, leave it as it is

File: src/app/modules/users/components/users-list/users-list.component.html [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+      <th class="col-4">Names</th>
+      <th class="col-4 text-center">Admin Role</th>
+    </tr>
+  </thead>
+  <app-loading-bar *ngIf="isLoading$ | async"></app-loading-bar>
+  <tbody *ngIf="(isLoading$ | async) === false">
+    <tr class="d-flex" *ngFor="let user of users">
+      <td class="col-sm-4">{{ user.email }}</td>
+      <td class="col-sm-4">{{ user.name }}</td>
+      <td class="col-sm-4 text-center custom-control custom-switch">
```
Comment (enriquezrene) [inline src/app/modules/users/components/users-list/users-list.component.html:14]: This is going to be confusing. Let's either make this column invisible or remove it completely.

Comment (PaulRC-ioet) [reply src/app/modules/users/components/users-list/users-list.component.html:14]: Ok, I will remove them, that option will enter in another ticket.

**Comment by PaulRC-ioet** [comment]
Captura de pantalla de 2020-11-23 14-32-56
this would be the final design
